File: python/ack_cost_sdk/costv2.py
"""
Cost V2 API服务模块
"""

from typing import Optional, Dict, List
from urllib.parse import urlencode
from dataclasses import dataclass, field
import json
from .client import Client
from .models import Properties, CostData, Response
from .errors import CostError, http_error_from_status_code


@dataclass
class CostV2Request:
    """Cost V2 API请求参数"""
    window: str                                    # 查询的持续时间
    filter: Optional[str] = None                  # 资源过滤条件
    step: Optional[str] = None                    # 时间分段
    aggregate: Optional[str] = None               # 聚合维度
    idle: Optional[bool] = None                   # 是否展示闲置成本
    share_idle: Optional[bool] = None             # 是否分摊闲置成本
    share_split: Optional[str] = None             # 闲置分摊策略
    idle_by_node: Optional[bool] = None           # 是否按节点维度聚合闲置成本
    format: Optional[str] = None                  # 成本导出格式


@dataclass
class CostV2Data(CostData):
    """Cost V2 成本数据"""
    pass


@dataclass
class CostV2Response(Response):
    """Cost V2 API响应"""
    data: List[Dict[str, CostV2Data]] = field(default_factory=list)


class CostV2Service:
    """Cost V2 API服务"""
    
    def __init__(self, client: Client):
        self.client = client
    
    def get_cost_v2(self, request: CostV2Request) -> CostV2Response:
        """
        查询Cost V2成本数据
        
        Args:
            request: CostV2Request请求对象
            
        Returns:
            CostV2Response: 响应对象
            
        Raises:
            CostError: 请求失败或解析错误
        """
        # 构建URL
        api_url = f"{self.client.config.api_server}/api/v1/namespaces/kube-system/services/ack-metrics-adapter-api-service:8080/proxy/v2/cost"
        
        # 添加查询参数
        params = {}
        params["window"] = request.window
        
        if request.filter:
            params["filter"] = request.filter
            
        if request.step:
            params["step"] = request.step
            
        if request.aggregate:
            params["aggregate"] = request.aggregate
            
        if request.idle is not None:
            params["idle"] = str(request.idle).lower()
            
        if request.share_idle is not None:
            params["shareIdle"] = str(request.share_idle).lower()
            
        if request.share_split:
            params["shareSplit"] = request.share_split
            
        if request.idle_by_node is not None:
            params["idleByNode"] = str(request.idle_by_node).lower()
            
        if request.format:
            params["format"] = request.format
        
        # 构建完整URL
        if params:
            api_url = f"{api_url}?{urlencode(params)}"
        
        try:
            # 发送请求
            response = self.client.do_request("GET", api_url)
            
            # 检查响应状态码
            if response.status_code != 200:
                raise http_error_from_status_code(response.status_code)
            
            # 解析响应
            raw_data = response.json()
            result = CostV2Response()
            
            # 转换数据格式
            for item in raw_data.get("data", []):
                data_dict = {}
                for key, value in item.items():
                    cost_data = CostV2Data()
                    cost_data.name = value.get("name", "")
                    
                    # 处理properties
                    props = value.get("properties", {})
                    cost_data.properties = Properties(
                        pod=props.get("pod"),
                        node=props.get("node"),
                        namespace=props.get("namespace"),
                        controller_kind=props.get("controllerKind"),
                        controller=props.get("controller"),
                        provider_id=props.get("providerID"),
                        labels=props.get("labels", {})
                    )
                    
                    cost_data.start = value.get("start", "")
                    cost_data.end = value.get("end", "")
                    cost_data.cpu_core_request_average = value.get("cpuCoreRequestAverage", 0.0)
                    cost_data.cpu_core_usage_average = value.get("cpuCoreUsageAverage", 0.0)
                    cost_data.ram_byte_request_average = value.get("ramByteRequestAverage", 0.0)
                    cost_data.ram_byte_usage_average = value.get("ramByteUsageAverage", 0.0)
                    cost_data.cost = value.get("cost", 0.0)
                    cost_data.cost_ratio = value.get("costRatio", 0.0)
                    cost_data.custom_cost = value.get("customCost", 0.0)
                    
                    data_dict[key] = cost_data
                result.data.append(data_dict)
            
            return result
            
        except json.JSONDecodeError as e:
            raise CostError("InternalError", "Failed to decode response", e)
        except Exception as e:
            if isinstance(e, CostError):
                raise e
            raise CostError("InternalError", "Failed to send request", e)

File: javascript/src/costv2.js
/**
 * Cost V2 API服务模块
 * @module costv2
 */

const { httpErrorFromStatusCode } = require('./errors');

/**
 * Cost V2 API请求参数类
 */
class CostV2Request {
    /**
     * 构造函数
     * @param {Object} options - 请求参数
     * @param {string} options.window - 查询的持续时间
     * @param {string} [options.filter] - 资源过滤条件
     * @param {string} [options.step] - 时间分段
     * @param {string} [options.aggregate] - 聚合维度
     * @param {boolean} [options.idle] - 是否展示闲置成本
     * @param {boolean} [options.shareIdle] - 是否分摊闲置成本
     * @param {string} [options.shareSplit] - 闲置分摊策略
     * @param {boolean} [options.idleByNode] - 是否按节点维度聚合闲置成本
     * @param {string} [options.format] - 成本导出格式
     */
    constructor(options) {
        this.window = options.window;
        this.filter = options.filter;
        this.step = options.step;
        this.aggregate = options.aggregate;
        this.idle = options.idle;
        this.shareIdle = options.shareIdle;
        this.shareSplit = options.shareSplit;
        this.idleByNode = options.idleByNode;
        this.format = options.format;
    }
}

/**
 * Kubernetes对象属性类
 */
class CostV2Properties {
    /**
     * 构造函数
     * @param {Object} options - 属性选项
     * @param {string} [options.pod] - Pod名称
     * @param {string} [options.node] - 节点名称
     * @param {string} [options.namespace] - 命名空间
     * @param {string} [options.controllerKind] - 控制器类型
     * @param {string} [options.controller] - 控制器名称
     * @param {string} [options.providerID] - 节点对应ECS实例ID
     * @param {Object} [options.labels] - Pod标签
     */
    constructor(options = {}) {
        this.pod = options.pod;
        this.node = options.node;
        this.namespace = options.namespace;
        this.controllerKind = options.controllerKind;
        this.controller = options.controller;
        this.providerID = options.providerID;
        this.labels = options.labels || {};
    }
}

/**
 * Cost V2 成本数据类
 */
class CostV2Data {
    /**
     * 构造函数
     * @param {Object} options - 数据选项
     * @param {string} [options.name] - 名称
     * @param {CostV2Properties} [options.properties] - 属性
     * @param {string} [options.start] - 开始时间
     * @param {string} [options.end] - 结束时间
     * @param {number} [options.cpuCoreRequestAverage] - 请求的平均CPU核心数
     * @param {number} [options.cpuCoreUsageAverage] - 使用的平均CPU核心数
     * @param {number} [options.ramByteRequestAverage] - 请求的平均内存量
     * @param {number} [options.ramByteUsageAverage] - 使用的平均内存量
     * @param {number} [options.cost] - 估算成本
     * @param {number} [options.costRatio] - 成本占比
     * @param {number} [options.customCost] - 自定义成本
     */
    constructor(options = {}) {
        this.name = options.name || "";
        this.properties = options.properties || new CostV2Properties();
        this.start = options.start || "";
        this.end = options.end || "";
        this.cpuCoreRequestAverage = options.cpuCoreRequestAverage || 0.0;
        this.cpuCoreUsageAverage = options.cpuCoreUsageAverage || 0.0;
        this.ramByteRequestAverage = options.ramByteRequestAverage || 0.0;
        this.ramByteUsageAverage = options.ramByteUsageAverage || 0.0;
        this.cost = options.cost || 0.0;
        this.costRatio = options.costRatio || 0.0;
        this.customCost = options.customCost || 0.0;
    }
}

/**
 * Cost V2 API响应类
 */
class CostV2Response {
    /**
     * 构造函数
     * @param {Object} options - 响应选项
     * @param {Array<Object>} [options.data] - 数据数组
     */
    constructor(options = {}) {
        this.data = options.data || [];
    }
}

/**
 * Cost V2 API服务类
 */
class CostV2Service {
    /**
     * 构造函数
     * @param {Client} client - SDK客户端
     */
    constructor(client) {
        this.client = client;
    }
    
    /**
     * 查询Cost V2成本数据
     * @param {CostV2Request} request - 请求参数
     * @returns {Promise<CostV2Response>} 响应数据
     */
    async getCostV2(request) {
        // 构建URL
        let apiURL = `${this.client.config.apiServer}/api/v1/namespaces/kube-system/services/ack-metrics-adapter-api-service:8080/proxy/v2/cost`;
        
        // 添加查询参数
        const params = new URLSearchParams();
        params.append('window', request.window);
        
        if (request.filter) {
            params.append('filter', request.filter);
        }
        
        if (request.step) {
            params.append('step', request.step);
        }
        
        if (request.aggregate) {
            params.append('aggregate', request.aggregate);
        }
        
        if (request.idle !== undefined) {
            params.append('idle', request.idle.toString());
        }
        
        if (request.shareIdle !== undefined) {
            params.append('shareIdle', request.shareIdle.toString());
        }
        
        if (request.shareSplit) {
            params.append('shareSplit', request.shareSplit);
        }
        
        if (request.idleByNode !== undefined) {
            params.append('idleByNode', request.idleByNode.toString());
        }
        
        if (request.format) {
            params.append('format', request.format);
        }
        
        // 构建完整URL
        if (params.toString()) {
            apiURL = `${apiURL}?${params.toString()}`;
        }
        
        try {
            // 发送请求
            const response = await this.client.doRequest({
                method: 'GET',
                url: apiURL
            });
            
            // 检查响应状态码
            if (response.status !== 200) {
                throw httpErrorFromStatusCode(response.status);
            }
            
            // 解析响应
            const rawData = response.data;
            const result = new CostV2Response();
            
            // 转换数据格式
            result.data = (rawData.data || []).map(item => {
                const dataDict = {};
                for (const [key, value] of Object.entries(item)) {
                    const costData = new CostV2Data({
                        name: value.name,
                        properties: new CostV2Properties({
                            pod: value.properties?.pod,
                            node: value.properties?.node,
                            namespace: value.properties?.namespace,
                            controllerKind: value.properties?.controllerKind,
                            controller: value.properties?.controller,
                            providerID: value.properties?.providerID,
                            labels: value.properties?.labels || {}
                        }),
                        start: value.start,
                        end: value.end,
                        cpuCoreRequestAverage: value.cpuCoreRequestAverage,
                        cpuCoreUsageAverage: value.cpuCoreUsageAverage,
                        ramByteRequestAverage: value.ramByteRequestAverage,
                        ramByteUsageAverage: value.ramByteUsageAverage,
                        cost: value.cost,
                        costRatio: value.costRatio,
                        customCost: value.customCost
                    });
                    dataDict[key] = costData;
                }
                return dataDict;
            });
            
            return result;
        } catch (error) {
            if (error instanceof httpErrorFromStatusCode().constructor) {
                throw error;
            }
            throw new Error(`Failed to send request: ${error.message}`);
        }
    }
}

module.exports = {
    CostV2Service,
    CostV2Request,
    CostV2Properties,
    CostV2Data,
    CostV2Response
};